daemon/CryptoImpl/crypto.cpp

```diff
@@ -1234,14 +1367,39 @@ void Daemon::ApiImpl::CryptoRequestQueue::handleFinishedRequest(
             }
             break;
         }
+        case AuthenticatedEncryptRequest: {
+            Result result = request->outParams.size()
+                    ? request->outParams.takeFirst().value<Result>()
+                    : Result(Result::UnknownError,
+                             QLatin1String("Unable to determine result of AuthenticatedEncryptRequest request"));
+            if (result.code() == Result::Pending) {
+                // shouldn't happen!
+                qCWarning(lcSailfishCryptoDaemon) << "EncryptRequest:" << request->requestId << "finished as pending!";
+                *completed = true;
+            } else {
+                QByteArray encrypted = request->outParams.size()
+                        ? request->outParams.takeFirst().toByteArray()
+                        : QByteArray();
+                QByteArray tag = request->outParams.size()
+                        ? request->outParams.takeFirst().toByteArray()
+                        : QByteArray();
+                request->connection.send(request->message.createReply() << QVariant::fromValue<Result>(result)
+                                                                        << QVariant::fromValue<QByteArray>(encrypted)
+                                                                        << QVariant::fromValue<QByteArray>(tag));
+                *completed = true;
+            }
+            break;
+        }
+        case AuthenticatedDecryptRequest:
+            // fall through, result/parameters are the same as for non-authenticated
```

Comment:
Are they?  Shouldn't it also produce a boolean "verified" out-parameter?

Reply:
According to https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption checking the result of EVP_DecryptFinal_ex() is sufficient, or have I missed something?

Reply:
@chriadam Could you clarify the purpose of the "verified" return value for finaliseCipherSession() in the plugins? I would have thought that returning an "error" result would be sufficient if the EVP_DecryptFinal_ex() operation fails, instead of requiring an additional "verified" value (assuming that the decrypted data should just be discarded if this verification fails).

Reply:
In GCM mode, the EVP_DecryptFinal_ex() value does not produce additional plaintext, instead it produces (in effect) a boolean "did the calculated tag match the provided tag" result.  The decrypted data (plaintext) should still be returned to the client (although in practice the client should not use that returned plaintext, since it is guaranteed to have been generated from corrupted ciphertext).

Reply:
Also, note that no error has occurred - GCM has worked successfully in this case - that is, it successfully completed and determined that the plaintext should not be trusted.

Reply:
Okay, I see. Would it make sense to return some other error, e.g. AuthenticationVerificationFailed or something instead of a 'Success' code, and then still return the decrypted plaintext?

Reply:
Do you think the Verify operation should produce a boolean "verified" outparam, or just return a VerificationFailed vs Success code in the result?  Whatever we decide, we should be consistent across all operations.

Personally, I'm partial to using explicit outparams, and using error codes only for actual errors (that is, something went wrong), but I'm willing to be convinced that that is less intuitive for users or otherwise suboptimal.